### PR TITLE
Downloading files in rust

### LINKED ISSUE #122: Downloading files from github
Add the downloading of the cache files and the churn file to the rust code, so it can be used from android and iOS.

### PR DESCRIPTION
**Pull request by ineiti**
This PR puts the downloading of the cache files to the rust part, so that both Android and iOS can rely on the same code.

Closes #122

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -14,23 +18,51 @@ use tor_dirmgr::Error;
...
+#[derive(PartialEq)]
+enum UpdateNeeded {
+    None,
+    Churn,
+    All,
+}
+
+/// Default directory cache download URL, provided by C4DT.
+pub const DIRECTORY_CACHE_C4DT: &str =
+    "https://github.com/c4dt/lightarti-directory/releases/latest/download/directory-cache.tgz";
```
Comment (PascalinDe) [inline src/client.rs:43]: while we're at it, can't we make this configurable?

Comment (ineiti) [reply src/client.rs:43]: Yes, you can call `new_with_url` and give the URL for the directory cache and the directory churn file.

Comment (ineiti) [reply src/client.rs:43]: Or would it be better if there is only one `new` which takes the two URLs? And then you have to add the URLs also in the example code?

Like:
- `new` takes the URLs
- `new_with_defaults` takes the URLs given for github?

File: tests/client.rs [hunk 2]
```diff
@@ -134,7 +131,7 @@ async fn test_directory_not_existing() {
     let root_cause = error.root_cause();
     assert_eq!(
         format!("{}", root_cause),
-        "Corrupt cache: directory cache does not exist"
+        "No such file or directory (os error 2)"
```
Comment (PascalinDe) [inline tests/client.rs:134]: this looks like a system error to me, but this test was meant to test the case that a missing directory cache is handled explicitely by the program (i.e. gives an exact error message to the user)

**Review by PascalinDe — CHANGES_REQUESTED** [review]

File: tests/client.rs (hunk 2 above)
Comment (ineiti) [reply tests/client.rs:134]: Ah, OK, now I understand better what you wrote in the chat :) I'll adapt the test to return the more informative message.

File: tests/client.rs [hunk 3]
```diff
@@ -134,7 +131,7 @@ async fn test_directory_not_existing() {
     let root_cause = error.root_cause();
     assert_eq!(
         format!("{}", root_cause),
-        "Corrupt cache: directory cache does not exist"
+        "Corrupt cache: cache-directory doesn't exist"
```
Comment (PascalinDe) [inline tests/client.rs:134]: if you insist on changing the message, you need to change this test as well: https://github.com/c4dt/lightarti-rest-android/blob/e34f0ee7e58f5ae962d88d6ef612f6ad04ebfd37/artiwrapper/src/androidTest/java/org/c4dt/artiwrapper/JniTest.java#L85

Comment (ineiti) [reply tests/client.rs:134]: Thanks - I thought of releasing a new version (should it even be a point version, like 0.5.0?) and then adjusting the android and iOS libraries.

https://github.com/c4dt/lightarti-rest-android/issues/32

Comment (PascalinDe) [reply tests/client.rs:134]: Oki :+1: also OK for the point version, since we moved functionality to here

**Review by PascalinDe — APPROVED** [review]

**Pull request merged**